modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,254 @@
+    -> std::future<ActionServerResponse<ReplyT>>;
+
+/// Request the action server to stop.
+[[nodiscard]] auto requestActionServerToStopExecution(Session& session,
+                                                      const TopicConfig& topic_config) -> bool;
+
+// TODO: add a function to get notified when the server is idle again.
+
+// ----------------------------------------------------------------------------------------------------------
+// --------- ActionServer Implementation --------------------------------------------------------------------
+// ----------------------------------------------------------------------------------------------------------
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+ActionServer<RequestT, StatusT, ReplyT>::ActionServer(SessionPtr session, TopicConfig topic_config,
+                                                      ActionTriggerCallback&& action_trigger_cb,
+                                                      ExecuteCallback&& execute_cb)
+  : session_(std::move((session)))
+  , topic_config_(std::move(topic_config))
+  , action_trigger_cb_(std::move(action_trigger_cb))
+  , execute_cb_(std::move(execute_cb))
+  , request_service_(std::make_unique<Service<RequestT, ActionServerRequestResponse>>(
+        session_, topic_config_, [this](const RequestT& request) { return onRequest(request); }))
+  , request_consumer_([this](const RequestT& request) { return execute(request); }, std::nullopt) {
+  request_consumer_.start();
+}
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+ActionServer<RequestT, StatusT, ReplyT>::~ActionServer() {
+  request_consumer_.stop();
+}
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+auto ActionServer<RequestT, StatusT, ReplyT>::onRequest(const RequestT& request)
+    -> ActionServerRequestResponse {
+  if (is_running_) {
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): server is already serving one request.",
+                              topic_config_.name);
+
+    return { .status = ActionServerRequestStatus::REJECTED_ALREADY_RUNNING };
+  }
+
+  try {
+    const auto response = action_trigger_cb_(request);
+    if (response != ActionServerTriggerStatus::SUCCESSFUL) {
+      return { .status = ActionServerRequestStatus::REJECTED_USER };
+    }
+  } catch (const std::exception& ex) {
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): request callback failed with exception: {}.",
+                              topic_config_.name, ex.what());
+    return { .status = ActionServerRequestStatus::INVALID };
+  }
+
+  if (!request_consumer_.queue().tryPush(request)) {
+    // NOTE: this should never happen as the queue is unbound.
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): failed to push the job in the queue. THIS SHOULD "
+                              "NOT HAPPEN CHECK THE CODE.",
```

Comment:
I would avoid all caps. A `NOTE: this should not happen ...` should be sufficient.